### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by geeeb95**
First linear models

### CODE REVIEW

File: src/module_3/module_3.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+from sklearn.linear_model import Ridge
+from sklearn.ensemble import GradientBoostingClassifier
+from sklearn.ensemble import RandomForestClassifier
+from sklearn import metrics
+from sklearn.preprocessing import normalize
+
+import matplotlib.pyplot as plt
+
+# %% Load data
+address = "D:\\Yo\\"
```
Comment (gsantam) [inline src/module_3/module_3.py:15]: In general, in a Python script ever add code outside a function (except global variable declarations). So create a function (for example `def train_model()`) and include the logic there. In order to make the script executable as a program (for example from the console) include `if __name__=='__main__` at the end ([check this out](https://medium.com/@mycodingmantras/what-does-if-name-main-mean-in-python-fa6b0460a62d#:~:text=In%20this%20example%2C%20the%20main,function%20to%20execute%20the%20code) for more details)

File: src/module_3/module_3.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+# Obtain the id's for the orders where 5 or more products where bought
+filter_df = df[df["outcome"]==1]
+filter_df = filter_df.groupby("order_id").filter(lambda x: len(x) > 4)
+
+# Filter the data so we retain only those ids
+dataset = df[df["order_id"].isin(filter_df["order_id"])]
+
+# Change date format to retain only year, month and day
+dataset["created_at"] = dataset["created_at"].str[:-9]
```
Comment (gsantam) [inline src/module_3/module_3.py:29]: I am not sure what the format of the original date is here, but normally is better to do first `pd.to_datetime(dataset["created_at"])` in order to cast the column to a date, and then you can just convert it to whatever string you want in a safe way

File: src/module_3/module_3.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+dataset["created_at"] = dataset["created_at"].str.replace("-","")
+pd.to_numeric(dataset["created_at"])
```
Comment (gsantam) [inline src/module_3/module_3.py:31]: this line is not doing anything. If you want to have some effect you need to do: `dataset["created_at"] = pd.to_numeric(dataset["created_at"])`

File: src/module_3/module_3.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+dataset["order_date"] = dataset["order_date"].str[:-9]
+dataset["order_date"] = dataset["order_date"].str.replace("-","")
+pd.to_numeric(dataset["order_date"])
```
Comment (gsantam) [inline src/module_3/module_3.py:34]: same as before

File: src/module_3/module_3.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+# Order by date
+dataset = dataset.sort_values(by=["order_date"])
+
+# %% Drop qualitative features
+print(dataset.columns)
+print(dataset.dtypes)
+dataset.drop(["product_type" , "vendor"], axis=1, inplace=True)
```
Comment (gsantam) [inline src/module_3/module_3.py:42]: I prefer to put it as `dataset.drop(columns = ["product_type" , "vendor"], inplace=True)` in order to be more explicit

File: src/module_3/module_3.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+# %% Divide the data in training, validation and test set
+def timesplit_by_order_id(data: pd.DataFrame, train_size: float, validation_test_size: float):
+    total_orders = data.order_id.unique()
+    train_num = round(total_orders.size*train_size)
```
Comment (gsantam) [inline src/module_3/module_3.py:47]: nit pick, I prefer to use `len(total_orders)` since it's the native `len` of Python

File: src/module_3/module_3.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    val_num = round(total_orders.size*validation_test_size)
+    test_num = val_num
+    gap = round(total_orders.size*(1-validation_test_size-train_size)/3)
+
+    train_orders=total_orders[0:train_num]
```
Comment (gsantam) [inline src/module_3/module_3.py:52]: to do the `train`, `test`, `val` split is probably a good idea to use https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

Comment (gsantam) [reply src/module_3/module_3.py:52]: Small comment: it's normally better and more elegant to do `train_orders=total_orders[:train_num]` i.e. avoid the 0

File: src/module_3/module_3.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    val_orders=total_orders[train_num+gap:train_num+gap+val_num]
+    test_orders=total_orders[train_num+2*gap+val_num:train_num+2*gap+val_num+test_num]
```
Comment (gsantam) [inline src/module_3/module_3.py:54]: this looks very complicated, I would do (supposing that I'm understanding correctly what are you trying to do) something like:
```
train_orders=total_orders[:train_num]
val_orders =total_orders[train_num:train_num+val_num]
test_orders =total_orders[train_num+val_num:]
```

Comment (gsantam) [reply src/module_3/module_3.py:54]: Also, very very important, if you are doing a random train / val / test split ALWAYS ensure that your dataset has been shuffled before (something like `total_orders = total_orders.sample(frac = 1)`) otherwise train / val / test may not have the same distribution

File: src/module_3/module_3.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+    train = dataset[dataset["order_id"].isin(train_orders)]
+    val = dataset[dataset["order_id"].isin(val_orders)]
+    test = dataset[dataset["order_id"].isin(test_orders)]
+
+    return train, val, test
+
+train_set, val_set, test_set = timesplit_by_order_id(dataset, 0.74, 0.10)
```
Comment (gsantam) [inline src/module_3/module_3.py:62]: Okey reading the name of the function is not clear to me if you are trying to do a random split or a split by time (I added [this](https://github.com/geeeb95/zrive-ds/pull/2/files#r1393410254) comment thinking that you wanted to do a random split) . If you do want to do a split by time I don't think is as easy as sorting the orders in chronological order and then taking the first chunk as train, the second as eval and the last as test. The reason is that you don't know how these orders are distributed over time: imagine that the first orders are very spread apart while the last ones (the most recent ones) are happening in a very small period of time. This would mean that, in a very extreme case, the orders of eval and test may happen even in the same day, while train ones may happen over several months, which doesn't seem like a great split. Instead, what one normally does is to split the time itself: for example, train would be all the orders happening from Jan to August (independently on how many orders are them), eval would be the orders from Sept to Oct and test the ones from November to December. One also normally does some EDA to understand how the orders are spread over time

File: src/module_3/module_3.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+X_train = train_set.drop(["outcome"], axis=1)
+y_train = train_set["outcome"]
+X_val = val_set.drop(["outcome"], axis=1)
+y_val = val_set["outcome"]
+X_test = test_set.drop(["outcome"], axis=1)
+y_test = test_set["outcome"]
+
+# %% Train linear model(I will try a couple of them)
```
Comment (gsantam) [inline src/module_3/module_3.py:71]: why the %%?

File: src/module_3/module_3.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+reg1 = linear_model.LinearRegression()
```
Comment (gsantam) [inline src/module_3/module_3.py:73]: what happens with the categorical features?

File: src/module_3/module_3.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+reg2 =linear_model.Ridge()
+reg2.fit(X_train, y_train)
+
+# %% Train non-linear models (tree based)
+
+tree1 = RandomForestClassifier(max_depth=2, random_state=0)
+tree1.fit(X_train, y_train)
+
+tree2 = GradientBoostingClassifier(n_estimators=100, learning_rate=0.3,
+max_depth=2, random_state=0)
```
Comment (gsantam) [inline src/module_3/module_3.py:85]: why this max_depth?

File: src/module_3/module_3.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+# First create an array with the baseline and predictions
+y=[]
+y.append(test_set["global_popularity"])
+y.append(reg1.predict(X_test))
+y.append(reg2.predict(X_test))
+y.append(tree1.predict(X_test))
+y.append(tree2.predict(X_test))
+
+#calculate precision, recall, FPR and TPR
+precision = [None]*len(y)
```
Comment (gsantam) [inline src/module_3/module_3.py:98]: why are you using these specific metrics? what do they tell you about the business problem and how do they translate to $?

**Review by gsantam — CHANGES_REQUESTED** [review]
Added some comments, specially about the evaluation. I think it would be a good idea to have done some of the work in a notebook to be able to explore the data, the features, the different metrics...
